### PR TITLE
AP-9806 # call API to determine payment amount

### PR DESCRIPTION
**Pull request by divporter**
## Requester Checklist

Please only check the items that you have actioned. Do not check items that are not applicable to your PR.

### Implementation

- [ ] Have you tested your implementation locally?
- [ ] If applicable, has an appropriate changelog entry been added? Do not include if you are fixing/changing something that is only in the current release.
- [ ] There are no warnings that have been suppressed unnecessarily
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
- [ ] Have all OneBlink dependency updates been completed (eg apps / apps-react / types etc).
- [ ] Have you ensured this change does not add unwanted dependencies?
- [ ] If this PR contains a refactor, have relevant Jira testing tasks added?
- [ ] Changes that will knowingly make the feature/bug incomplete have been commented with TODO and a description of what needs to be done to finish the feature/bug
- [ ] Any changes made to public APIs have been reflected in the documentation
- [ ] Have you isolated business logic where possible to allow for unit testing?

### Logging and Debugging

- [ ] Are the error messages, if any, informative?
- [ ] Are there enough log events and are they written in a way that allows for easy debugging?
- [ ] "Debugging" code removed
- [ ] Front-end: No erroneous `Console.WriteLines`

### Readability

- [ ] All class, variable, property and method modifiers are provided with the smallest scope possible
- [ ] New files, variables and functions are descriptive/comprehensible and named consistently.
- [ ] There is no dead code (unreachable code)
- [ ] There is no usage of [magic numbers](<https://en.wikipedia.org/wiki/Magic_number_(programming)>)
- [ ] There is no commented out code.
- [ ] In hard-to-understand areas, comments exist and describe rationale or reasons for decisions in code

### Security

- [ ] All personal data inputs are checked (for the correct type, length/size, format, and range).
- [ ] No sensitive information is logged or visible in a stacktrace
- [ ] Are authorization and authentication handled correctly?
- [ ] Is (user) input validated, sanitized, and escaped to prevent security attacks such as cross-site scripting or SQL injection?
- [ ] Is data retrieved from external APIs or libraries checked for security issues?
- [ ] Do API endpoints return appropriate status codes

## Reviewer Guide

- It is important that you understand the purpose of the PR.
- You are encouraged to engage with the requestor if you do not understand any of the proposed code changes/additions/deletions.
- Do you, the reviewer, understand what the code does? Do you think a specific expert, like a security expert or a usability expert, should look over the code before it can be accepted?
- Is a framework, API, library, or service used that should not be used? Are there alternatives you could recommend?
- Are there existing hooks/components/functions in the same code base that could be utilised?
- When reviewing tests, attempt to identify missing edge cases that may be relevant to the proposed implementation
- Ensure you check for:
  - Security
  - Scalability
  - Performance
  - Maintainability

---

> [!NOTE]
> **High Risk**
> Changes the end-to-end payment initiation flow and amount authority (now API-driven), including retry and zero-amount edge cases, in a security- and revenue-sensitive area.
> 
> **Overview**
> **Payment configuration is now created server-side** instead of the client building provider-specific `POST /forms/:id/...` payloads. `createPaymentConfiguration` calls `POST /form-submission-meta/${submissionId}/payment` with redirect URLs and receives `hostedFormUrl`, `amount`, and `paymentSubmissionEvent` (or no body when payment is not required).
> 
> `handlePaymentSubmissionEvent` no longer takes a client-computed amount or event; it relies on that API response and can return `null`, which clears stored payment state and skips the hosted form. **Try Again** on the payment receipt treats a `null` response as “no payment needed” and runs the success (`Done`) path.
> 
> Client-side `checkForPaymentSubmissionEvent` now only signals whether a payment event exists (amount &gt; 0) for gating offline/upload flows; it no longer supplies the payable amount. **`preparePaymentConfiguration` is removed** from all payment providers and from the `PaymentProvider` type—providers are used only for transaction verification/receipts. Submit and scheduling paths were updated to match the slimmer payment-event check and `handlePaymentSubmissionEvent` signature.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit ce891de96b749c69f3937f38984ffa584a9f1e42. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/apps/services/api/payment.ts [hunk 1]
```diff
@@ -307,22 +301,29 @@ const verifyPaymentTransaction = <T>(
...
 }
 
-const acknowledgeCPPayTransaction = async (
-  formId: number,
-  payload: unknown,
-): Promise<void> => {
-  const url = `${tenants.current.apiOrigin}/forms/${formId}/cp-pay-acknowledge`
-  console.log('Attempting to acknowledge CP Pay transaction', url)
-  await postRequest(url, payload).catch((error) => {
+async function createPaymentConfiguration(
```
Comment (mymattcarroll) [inline src/apps/services/api/payment.ts]: Sorry to be jerk but would you mind moving this to where `generatePaymentConfiguration()` was? Would make this much easier to review.

Comment (divporter) [reply src/apps/services/api/payment.ts]: I'll take a look.

File: src/apps/payment-service.ts [hunk 2]
```diff
@@ -143,10 +138,11 @@ export function checkForPaymentSubmissionEvent(
     startOfDay,
     endOfDay,
   })
-  if (result) {
-    console.log('Form has a payment submission event with amount', result)
+  if (result && result.amount > 0) {
```
Comment (mymattcarroll) [inline src/apps/payment-service.ts:141]: `result.amount > 0` I dont mind this logic being here but just want to double check: `result` would be `undefined` if amount was calculated as 0, right?

File: src/PaymentReceipt.tsx [hunk 3]
```diff
@@ -199,16 +198,22 @@ function PaymentReceipt({
...
         paymentFormUrl: submissionResult.payment.paymentFormUrl,
       })
-      await submissionService.executePostSubmissionAction({ ...submissionResult, payment }, {
-        onRedirectToRelativeUrl: (url) => history.push(url),
-        onRedirectToAbsoluteUrl: (url) => window.location.assign(url),
-      })
+      // The API has decided that no payment is required on try again
+      if (!payment) {
+        // Do the success path
+        handleDone()
```
Comment (mymattcarroll) [inline src/PaymentReceipt.tsx:208]: Should there be a `return` after this?

File: src/apps/payment-service.ts (hunk 2 above)
Comment (divporter) [reply src/apps/payment-service.ts:141]: Yes core will return undefined if amount is zero. Interestingly not if it's a negative. I wonder if we should fix that, or just always return amount.

File: src/apps/services/api/payment.ts (hunk 1 above)
Comment (divporter) [reply src/apps/services/api/payment.ts]: Ah yes I get what you're saying. On it.

File: src/apps/payment-service.ts (hunk 2 above)
Comment (mymattcarroll) [reply src/apps/payment-service.ts:141]: 🤔 If production currently supports negatives, I vote we leave it supporting them.

Comment (divporter) [reply src/apps/payment-service.ts:141]: Agreed.

**Review by mymattcarroll — APPROVED** [review]
